doc/cprover-manual/contracts-assigns.md

```diff
@@ -1,75 +1,373 @@
+
+#define MAX_SIZE 10
+
+struct vec_t {
+  size_t size;
+  char *data;
+};
+
+// Allocates a vect_t struct together with its data and a hidden byte
+// in a same object.
+struct vec_t *vec_alloc(size_t size) {
+
+  if(size > MAX_SIZE)
+    size = MAX_SIZE;
+
+  // allocate the struct + data + 1 extra hidden byte
+  struct vec_t *vec = malloc(sizeof(struct vec_t) + size + 1);
+  if (vec) {
+    vec->size = size;
+    vec->data = ((char *)vec) + sizeof(struct vec_t);
+  }
+  return vec;
+}
+
+// Clear the vec->data array
+void vec_clear(struct vec_t *vec)
+  __CPROVER_assigns(
+    vec && vec->data: __CPROVER_object_from(vec->data))
+{
+  if (!vec)
+    return;
+
+  vec->size = vec->size; // FAILURE
+
+  for (size_t i = 0; i < vec->size; i++)
+    vec->data[i] = 0; // SUCCESS
+
+  char *hidden_byte = ((char *)vec + sizeof(*vec) + vec->size);
+  *hidden_byte = 0; // SUCCESS
+}
+
+// Proof harness
+int main() {
+  size_t size;
+  struct vec_t *vec = vec_alloc(size);
+  vec_clear(vec);
+}
+```
+---
+
+```c
+void __CPROVER_object_whole(void *ptr);
 ```
 
-### Semantics
+Given a pointer `ptr` pointing into some object (possibly at some non-zero offset),
+`__CPROVER_object_whole(ptr)` specifies that the range of bytes of size
+`__CPROVER_OBJECT_SIZE(ptr)` starting at address
+`ptr - __CPROVER_POINTER_OFFSET(ptr)` is assignable:
 
-The semantics of an _assigns_ clause of a given function can be understood
-in two contexts: enforcement and replacement.
+If the pointer has a positive offset into some object, the range includes
+bytes that are in the object before the address pointed to by `ptr`.
+Revisiting our example one last time, changing the target to
+`__CPROVER_object_whole(vec->data)` allows the function (perhaps mistakenly)
+to assign to `vec->size`, the whole array of size `vec->size` pointed to by
+`vec->data` and the hidden byte.
 
-#### Enforcement
+```c
+#include <stdlib.h>
+
+#define MAX_SIZE 10
+
+struct vec_t {
+  size_t size;
+  char *data;
+};
+
+// Allocates a vect_t struct together with its data and a hidden byte
+// in a same object.
+struct vec_t *vec_alloc(size_t size) {
+
+  if(size > MAX_SIZE)
+    size = MAX_SIZE;
+
+  // allocate the struct + data + 1 extra hidden byte
+  struct vec_t *vec = malloc(sizeof(struct vec_t) + size + 1);
+  if (vec) {
+    vec->size = size;
+    vec->data = ((char *)vec) + sizeof(struct vec_t);
+  }
+  return vec;
+}
+
+// Clear the vec->data array
+void vec_clear(struct vec_t *vec)
+  __CPROVER_assigns(
+    vec && vec->data: __CPROVER_object_whole(vec->data))
+{
+  if (!vec)
+    return;
+
+  vec->size = vec->size; // SUCCESS
+
+  for (size_t i = 0; i < vec->size; i++)
+    vec->data[i] = 0; // SUCCESS
+
+  char *hidden_byte = ((char *)vec + sizeof(*vec) + vec->size);
+  *hidden_byte = 0; // SUCCESS
+}
 
-In order to determine whether an _assigns_ clause is a sound abstraction of
-the write set of a function *f*, the body of the function is instrumented with
+// Proof harness
+int main() {
```

Comment:
Are these examples all in the regression somewhere?  If so, is there a way we could link to them?

Reply:
No these are not in the regression test suite. I wouldn't know how to link the manual to the regression tests.